blog/_posts/2019-07-14-multithreading.md

```diff
@@ -0,0 +1,464 @@
+explicitly a multithreaded I/O library and so doesn't support concurrent use from
+multiple threads out of the box.
+We decided to protect access to libuv structures with a lock, and then allow any thread
+(one at a time) to run the event loop.
+When another thread needs the event loop thread to wake up, it issues an async signal.
+This can happen for multiple reasons, including another thread scheduling new work,
+or another thread needing to run garbage collection.
+
+### Task migration
+
+In general, a task might start running on one thread, block for a while, and then
+restart on another.
+This changes fundamental assumptions about when thread-local values can change.
+Internally, Julia code uses thread-local variables *constantly*, for example
+every time you allocate memory.
+We have yet to begin all the changes needed to support migration, so for now
+a task must always run on the thread it started running on (though, of course,
+it can start running on any thread).
+To support this, we have a concept of "sticky" tasks that must run on a given
+thread, and per-thread queues for running tasks associated with each thread.
+
+### Sleeping idle threads
+
+When there aren't enough tasks to keep all threads busy, some need to block to avoid
+using 100% of all CPUs all the time.
+This is a tricky synchronization problem, since some threads might be scheduling new work
+while other threads are deciding to block.
+
+### Where does the scheduler run?
+
+When a task blocks, we need to call the scheduler to pick another task to run.
+What stack does that code use to run?
+It's possible to have a dedicated scheduler task, but we felt there would be
+less overhead if we allowed the scheduler code to run in the context of the
+recently-blocked task.
+That works fine, but it means a task can exist in a strange intermediate state
+where it is considered not to be running, and yet is in fact running the
+scheduler.
+In particular, we need to make sure no other thread sees that task and thinks
+"oh, there's a task I can run", causing it to scribble on the scheduler's
+stack.
+
+### Classic bugs
+
+While trying to get this new functionality working, we encountered several
+maddeningly difficult bugs.
+My hands-down favorite was a mysterious hang on Windows that was fixed
+by literally [flipping a single bit][].
+
+
+## Looking forward
+
+While we are excited about this milestone, a lot of work remains.
+Here are some of the points we hope to focus on to further develop
+our threading capabilities:
+
+* Performance work on task switch and I/O latency.
+* Adding parallelism to the standard library. Many common operations like sorting and
+  array broadcasting can now use multiple threads internally.
+* Consider allowing task migration.
+* Improved debugging tools.
+* Explore API extensions like cancel points.
+* Provide alternate schedulers.
+
+
+## Acknowledgements
+
+We would like to gratefully acknowledge funding support from Intel and relational.ai
```

Comment:
```suggestion
We would like to gratefully acknowledge funding support from Intel and relationalAI
```

Reply:
Are they ok with capitalizing it as "RelationalAI"?